### PR TITLE
Use querysets from the class not from an instance

### LINKED ISSUE #4350: MultipleObjectsReturned in core.views.serve.serve.docs
We get this exception while trying to serve some docs:

```
MultipleObjectsReturned: get() returned more than one Version -- it returned 2!
  File "django/core/handlers/base.py", line 149, in get_response
    response = self.process_exception_by_middleware(e, request)
  File "django/core/handlers/base.py", line 147, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "readthedocs/core/views/serve.py", line 96, in inner_view
    return view_func(request, project=project, *args, **kwargs)
  File "readthedocs/core/views/serve.py", line 74, in inner_view
    return view_func(request, subproject=subproject, *args, **kwargs)
  File "readthedocs/core/views/serve.py", line 156, in serve_docs
    version = project.versions.public(request.user).get(slug=version_slug)
  File "django/db/models/query.py", line 391, in get
    (self.model._meta.object_name, num)
```

The view code looks legit at a first look but there's a huge side effect of the user filtering in the public() method. It does not filter the projects by user but it adds to the queryset all the other user projects which is not what we need here:

```
# project of the version we are looking for
>>> versione.project
<Project: Documentazione API anagrafe del Comune di Tre Civette sul Comò>
# this queryset is fine
>>> versione.project.versions.all()
[<Version: Version master of Documentazione API anagrafe del Comune di Tre Civette sul Comò (861)>, <Version: Version bozza of Documentazione API anagrafe del Comune di Tre Civette sul Comò (860)>]
# but here we have other projects' version too
>>> versione.project.versions.public(user)
[<Version: Version stabile of design-docs (859)>, <Version: Version master of Documentazione API anagrafe del Comune di Tre Civette sul Comò (861)>, <Version: Version bozza of design-docs (852)>, <Version: Version bozza of Documentazione API anagrafe del Comune di Tre Civette sul Comò (860)>]

```

### PR DESCRIPTION
**Pull request by stsewd**
When filtering using `public` and using a user,
the queryset hit this

https://github.com/rtfd/readthedocs.org/blob/45df7fd0da44be9eab3c0cb2888f6a9a15421fc5/readthedocs/builds/querysets.py#L22-L24

When the user is authenticated, we call to `get_objects_for_user`
which gets all the versions from all the user's projects.
Overriding any previous filter (`project.versions` in this case)

We don't see this in production because we serve from another domain.
And we don't see this in the corporate site because we override the
serve_docs view.

Fix #4350
Closes #4356

### CODE REVIEW

**Comment by stsewd** [comment]
Writing a test now

File: readthedocs/api/v2/views/footer_views.py [hunk 1]
```diff
@@ -25,7 +25,7 @@ def get_version_compare_data(project, base_version=None):
     :param base_version: We assert whether or not the base_version is also the
                          highest version in the resulting "is_highest" value.
     """
-    versions_qs = project.versions.public().filter(active=True)
+    versions_qs = Version.objects.public(project=project)
```
Comment (stsewd) [inline readthedocs/api/v2/views/footer_views.py]: filter by active is the default in .org and .com querysets

File: readthedocs/rtd_tests/tests/test_doc_serving.py [hunk 2]
```diff
@@ -293,3 +293,39 @@ def test_sitemap_xml(self):
...
+
+    @override_settings(
+        PYTHON_MEDIA=True,
+        USE_SUBDOMAIN=False,
+    )
+    @patch(
+        'readthedocs.core.views.serve._serve_symlink_docs',
+        new=mock.MagicMock(return_value=HttpResponse(content_type='text/html')),
+    )
+    def test_user_with_multiple_projects_serve_from_same_domain(self):
```
Comment (stsewd) [inline readthedocs/rtd_tests/tests/test_doc_serving.py]: This test case fails on master :)

**Comment by xrmx** [comment]
@stsewd any chance you can check this pr pass the tests of #4356 please?

**Comment by humitos** [comment]
I don't fully follow the problem and the solution.

What are the steps to reproduce it?

The solution seems to implicitly say that "project.versions." can't be used, is that correct?

**Comment by stsewd** [comment]
> The solution seems to implicitly say that "project.versions." can't be used, is that correct?

Yes.

Steps to reproduce:

- Create two projects with a common branch name (`latest`)
- The projects need to share the same owner
- Build the version of both projects
- While logged in (make sure to not have old cookies by logging out first)
- View the docs of one -- it returns the error (get returned 2 objects)
- (you need to use the same domain from the web app)

The problem is that when the user is authenticated, `get_objects_for_user` is called, which returns all the versions of all projects.

We filter this before, but all versions get added later anyway, so that breaks the unique slug on the queryset.

**Comment by stsewd** [comment]
> @stsewd any chance you can check this pr pass the tests of #4356 please?

It should pass, but it could break when using private versions and other privacy related stuff.

**Comment by ericholscher** [comment]
Should we add the tests from #4356 here, so we can close that one too?

**Comment by stsewd** [comment]
@ericholscher the test that was added here replace that test, and the test for admin access was added because the other PR drops the `public` from the queryset.

**Comment by ericholscher** [comment]
Sounds good 👍

**Pull request merged**